### PR TITLE
Selection model ported from napari

### PR DESCRIPTION
**Pull request by alisterburt**
I had a go at porting over the selection model and its test from napari - if you're happy with this I can go ahead and follow up with the `SelectableEventedList` 

I removed the pydantic validator and json encoder but will need to add those back if we want to have `EventedModel` here too

### CODE REVIEW

**Comment by tlambert03** [comment]
Wow! Blazing fast! :) 
Thanks, will have a close look soon

**Review by tlambert03 — COMMENTED** [review]
damn it, just came to check on this and realized all my reviews are still "pending" :joy:  guess I never hit enter.  

Anyway, need to make a few changes, mostly my own fault.  let me know if they make sense and happy to chat about it

File: psygnal/containers/_selection.py [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+        emitted when the current item has changed. (Private event)
+    """
+
+    events: SelectionEvents  # pragma: no cover
+
+    def __init__(self, data: Iterable[_T] = ()):
+        self._active: Optional[_T] = None
+        self._current_: Optional[_T] = None
+        super().__init__(iterable=data)
+        self.events = SelectionEvents()
```
[the remark this answers is not in the document]
Comment (alisterburt) [reply psygnal/containers/_selection.py:70]: got it, thanks for the clear walkthrough, will have a go and reach out if I get stuck!

**Comment by alisterburt** [comment]
set aside a few hours to fix this up and do battle with mypy/codecov, had fun and learnt a ton! Thanks again for the pointers

Should be ready for review - tried to fix some of the [flake errors](https://results.pre-commit.ci/run/github/383130468/1645914376.MzbhvPkYTWe7MWtCwzqf4A) here too

The biggest change is that `SelectionEvents.changed` became `SelectionEvents.items_changed` so we could continue leveraging the change event there to update the active selection member - if you don't like this I can probably find a way around it :)

I'm not sure cov will pass, I touched some files that weren't previously at 100% but all new stuff is at 100%

```
(psygnal) ➜  psygnal git:(selection) ✗ coverage report
Name                                     Stmts   Miss  Cover
------------------------------------------------------------
psygnal/__init__.py                         27      0   100%
psygnal/_group.py                           71     35    51%
psygnal/_signal.py                         361    274    24%
psygnal/_version.py                          2      0   100%
tests/containers/test_evented_list.py      188      4    98%
tests/containers/test_evented_proxy.py     117      0   100%
tests/containers/test_evented_set.py        67      0   100%
tests/containers/test_selection.py          73      0   100%
tests/test_bench.py                         31     26    16%
tests/test_group.py                         81      0   100%
tests/test_psygnal.py                      459    164    64%
tests/test_qt_compat.py                     59      0   100%
tests/test_utils.py                         21      0   100%
------------------------------------------------------------
TOTAL                                     1557    503    68%
```

**Review by tlambert03 — COMMENTED** [review]
Thanks for this @alisterburt!  I ended up just ignoring those new TC002 flake8 errors globally (don't really like that one).

I made some very minor tweaks directly to this PR and skipped a few strange coverage issues (it's not detecting coverage of decorators... but also, it doesn't seem to be merging reports correctly.  I gave up)

Thanks again and LGTM!

**Review by tlambert03 — APPROVED** [review]

**Comment by alisterburt** [comment]
Thanks for doing so much cleanup here @tlambert03 - I should've subclassed `SetEvents` myself :)

**Pull request merged**